packages/playwright-test/src/reporters/teleEmitter.ts

```diff
@@ -0,0 +1,221 @@
+      rootDir: config.rootDir,
+      configFile: config.configFile,
+    };
+  }
+
+  private _serializeProject(suite: Suite): JsonProject {
+    const project = suite.project()!;
+    const report: JsonProject = {
+      id: (project as any)._internal.id,
+      metadata: project.metadata,
+      name: project.name,
+      outputDir: project.outputDir,
+      repeatEach: project.repeatEach,
+      retries: project.retries,
+      testDir: project.testDir,
+      testIgnore: serializePatterns(project.testIgnore),
+      testMatch: serializePatterns(project.testMatch),
+      timeout: project.timeout,
+      suites: suite.suites.map(fileSuite => {
+        return this._serializeSuite(fileSuite);
+      }),
+      grep: serializePatterns(project.grep),
+      grepInvert: serializePatterns(project.grepInvert || []),
+      dependencies: project.dependencies,
+      snapshotDir: project.snapshotDir,
+    };
+    return report;
+  }
+
+  private _serializeSuite(suite: Suite): JsonSuite {
+    const location = this._relativeLocation(suite.location);
+    const result = {
+      type: suite._type,
+      title: suite.title,
+      fileId: (suite as any)._fileId,
+      location,
+      suites: suite.suites.map(s => this._serializeSuite(s)),
+      tests: suite.tests.map(t => this._serializeTest(t)),
+    };
+    return result;
+  }
+
+  private _serializeTest(test: TestCase): JsonTestCase {
+    return {
+      testId: test.id,
+      title: test.title,
+      location: this._relativeLocation(test.location)!,
+      expectedStatus: test.expectedStatus,
+      timeout: test.timeout,
+      annotations: test.annotations,
+      retries: test.retries,
+    };
+  }
+
+  private _serializeResultStart(result: TestResult): JsonTestResultStart {
+    const id = createGuid();
+    (result as any)[idSymbol] = id;
+    return {
+      id,
+      retry: result.retry,
+      workerIndex: result.workerIndex,
+      parallelIndex: result.parallelIndex,
+      startTime: result.startTime.toISOString(),
+    };
+  }
+
+  private _serializeResultEnd(result: TestResult): JsonTestResultEnd {
+    return {
+      id: (result as any)[idSymbol],
+      duration: result.duration,
+      status: result.status,
+      errors: result.errors,
+      attachments: result.attachments,
+    };
+  }
+
+  private _serializeStepStart(step: TestStep): JsonTestStepStart {
+    return {
+      id: (step as any)[idSymbol],
+      title: step.title,
+      category: step.category,
+      startTime: step.startTime.toISOString(),
+      location: this._relativeLocation(step.location),
+    };
+  }
+
+  private _serializeStepEnd(step: TestStep): JsonTestStepEnd {
+    return {
+      id: (step as any)[idSymbol],
+      duration: step.duration,
+      error: step.error,
+    };
+  }
+
+  private _relativeLocation(location: Location | undefined): Location | undefined {
+    if (!location)
+      return undefined;
+    const file = toPosixPath(path.relative(this.config.rootDir, location.file));
```

Comment:
I am not sure we want `toPosixPath` here, because we will use `file` to access file system later.